bots/flakes-refresh

```diff
@@ -0,0 +1,166 @@
+
+# Here we're looking for the count of a specific field/value in the record
+def count(record, field, only):
+    values = record.get(field, [])
+    for value, count in values:
+        if value != only:
+            continue
+        return count
+    return 0
+
+def run(api, verbose=False, dry=False):
+    open_issues = api.issues(labels=[ "flake" ])
+    create_count = NUMBER_OPEN_ISSUES - len(open_issues)
+
+    if create_count <= 0:
+        return 0
+
+    if verbose:
+        sys.stderr.write("Going to create %s new flake issue(s)\n" % create_count)
+
+    host = os.environ.get("COCKPIT_LEARN_SERVICE_HOST")
+    if not host:
+        raise RuntimeError("no learn host environment variable: $COCKPIT_LEARN_SERVICE_HOST")
+    port = os.environ.get("COCKPIT_LEARN_SERVICE_PORT", "8080")
+    url = "http://{0}:{1}/active/statistics.jsonl".format(host, port)
+
+    statistics = [ ]
+
+    # Retrieve the URL
+    try:
+        req = urllib.request.Request(url)
+        with urllib.request.urlopen(req) as f:
+            for line in f.readlines():
+                try:
+                    record = json.loads(line.decode('utf-8'))
+                    statistics.append(record)
+                except ValueError as ex:
+                    sys.stderr.write("{0}: {1}\n".format(url, ex))
+    except (ConnectionResetError, urllib.error.URLError, socket.gaierror) as ex:
+        sys.stderr.write("{0}: {1}\n".format(url, ex))
+        return False
+
+    tests = { }
+
+    for record in statistics:
+        test = value(record, "test")
+        context = value(record, "context")
+        status = value(record, "status")
+        tracker = value(record, "tracker")
+
+        # There should only be one of all of these values
+        if test is None or status is None:
+            continue
+
+        # Flaky tests only score on those that fail and are not tracked
+        if status == "failure" and not tracker:
+            merged = count(record, "merged", True)
+            not_merged = count(record, "merged", False)
+            null_merged = count(record, "merged", None)
+            total = merged + not_merged + null_merged
+
+            # And the key is that they were merged anyway
+            if total > 10:
+                failures = tests.get(test)
+                if not failures:
+                    tests[test] = failures = [ ]
+                failures.append((merged / total, context, record))
+
+    scores = [ ]
+
+    for n, t in tests.items():
+        scores.append((sum(map(lambda f: f[0], t))/len(t), n, t))
+
+    closed_issues = api.issues(labels=["flake"], state="closed", since=(time.time() - (WINDOW_DAYS * 86400)))
+
+    def find_in_issues(issues, name):
+        for issue in issues:
+            if name in issue["title"]:
+                return True
+        return False
+
+    scores.sort(reverse=True)
+    for score, name, failures in scores:
+        if find_in_issues(open_issues, name) or find_in_issues(closed_issues, name):
```

Comment:
So that means if a test that we fixed becomes flaky again, we would ignore it for 90 days. (Seems okay, and we can tune this, just thinking aloud)

Reply:
Yes.